deepdrivemd/deepdrivemd_stream.py

```diff
@@ -0,0 +1,253 @@
+        cfg.task_config.dump_yaml(cfg_path)
+        task = generate_task(cfg)
+        task.arguments += ["-c", cfg_path.as_posix()]
+        stage.add_tasks(task)
+
+        return stage
+
+    def generate_agent_stage(self) -> Stage:
+        stage = Stage()
+        stage.name = self.AGENT_STAGE_NAME
+        cfg = self.cfg.agent_stage
+        stage_api = self.api.agent_stage
+
+        task_idx = 0
+        output_path = stage_api.task_dir(self.stage_idx, task_idx, mkdir=True)
+        assert output_path is not None
+
+        # Update base parameters
+        cfg.task_config.experiment_directory = self.cfg.experiment_directory
+        cfg.task_config.stage_idx = self.stage_idx
+        cfg.task_config.task_idx = task_idx
+        cfg.task_config.node_local_path = self.cfg.node_local_path
+        cfg.task_config.output_path = output_path
+
+        # Write yaml configuration
+        cfg_path = stage_api.config_path(self.stage_idx, task_idx)
+        cfg.task_config.dump_yaml(cfg_path)
+        task = generate_task(cfg)
+        task.arguments += ["-c", cfg_path.as_posix()]
+        stage.add_tasks(task)
+
+        return stage
+
+
+if __name__ == "__main__":
+
+    args = parse_args()
+    cfg = StreamingExperimentConfig.from_yaml(args.config)
+    cfg.config_directory = os.path.dirname(os.path.abspath(args.config))
+    print("config_directory = ", cfg.config_directory)
+
+    cfg.adios_xml_sim = f"{cfg.config_directory}/adios_sim.xml"
+    cfg.adios_xml_agg = f"{cfg.config_directory}/adios_agg.xml"
+    cfg.agent_stage.task_config.adios_xml_agg = cfg.adios_xml_agg
+    cfg.aggregation_stage.task_config.adios_xml_agg = cfg.adios_xml_agg
+    cfg.machine_learning_stage.task_config.adios_xml_agg = cfg.adios_xml_agg
+    cfg.molecular_dynamics_stage.task_config.adios_xml_sim = cfg.adios_xml_sim
+
+    reporter = ru.Reporter(name="radical.entk")
+    reporter.title(cfg.title)
+
+    # Create Application Manager
+    try:
+        appman = AppManager(
+            hostname=os.environ["RMQ_HOSTNAME"],
+            port=int(os.environ["RMQ_PORT"]),
+            username=os.environ["RMQ_USERNAME"],
+            password=os.environ["RMQ_PASSWORD"],
+        )
+    except KeyError:
+        raise ValueError(
+            "Invalid RMQ environment. Please see README.md for configuring environment."
+        )
+
+    # Calculate total number of nodes required. Assumes 1 MD job per GPU
+    # TODO: fix this assumption for NAMD
+    num_full_nodes, extra_gpus = divmod(
+        cfg.molecular_dynamics_stage.num_tasks, cfg.gpus_per_node
+    )
+    extra_node = int(extra_gpus > 0)
+    num_nodes = max(1, num_full_nodes + extra_node)
+
+    appman.resource_desc = {
+        "resource": cfg.resource,
+        "queue": cfg.queue,
+        "schema": cfg.schema_,
+        "walltime": cfg.walltime_min,
+        "project": cfg.project,
+        "cpus": cfg.cpus_per_node * cfg.hardware_threads_per_cpu * num_nodes + 8,
```

Comment:
Why is there a +8 and +2 here?

Reply:
compute_number_of_nodes() is introduced to compute number of nodes more accurately.